core/common/src/main/java/alluxio/collections/LockCache.java

```diff
@@ -0,0 +1,239 @@
+        }
+      } finally {
+        mEvictLock.unlock();
+      }
+    }
+  }
+
+  /**
+   * If the key has an associated lock in the cache, return a LockResource for that key and
+   * the associated lock using the specified lockmode.  If the key has no entry in the cache,
+   * create a lock and return a LockResource for that lock.
+   *
+   * @param key the key to look up the cache
+   * @param mode lockMode to acquire
+   * @return the value contained in the cache, if it is already in cache,
+   *         otherwise generate an entry based on the loader
+   */
+  public LockResource get(K key, LockMode mode) {
+    ValNode valNode = getValNode(key);
+    ReentrantReadWriteLock lock = valNode.mValue;
+    switch (mode) {
+      case READ:
+        return new RefCountLockResource(lock.readLock(), valNode.mRefCount);
+      case WRITE:
+        return new RefCountLockResource(lock.writeLock(), valNode.mRefCount);
+      default:
+        throw new IllegalStateException("Unknown lock mode: " + mode);
+    }
+  }
+
+  /**
+   * Get the raw readwrite lock from the cache.
+   *
+   * @param key key to look up the value
+   * @return the lock associated with the key
+   */
+  @VisibleForTesting
+  public ReentrantReadWriteLock getRawReadWriteLock(K key) {
+    return mCache.getOrDefault(key, new ValNode(new ReentrantReadWriteLock())).mValue;
+  }
+
+  private ValNode getValNode(K key) {
+    Preconditions.checkNotNull(key, "key can not be null");
+    // oldCacheEntry keeps track of the last cache entry that was in the process of being removed.
+    // we do not need to do anything if we did not get a new CacheEntry
+    ValNode oldCacheEntry = null;
+    ValNode cacheEntry;
+    while (true) {
+      // repeat until we get a cacheEntry that is not in the process of being removed.
+      cacheEntry = mCache.compute(key, (k, v) -> {
+        if (v != null) {
+          v.mRefCount.incrementAndGet();
+          v.mIsAccessed = true;
+          return v;
+        }
+        // new cache entry
+        if (mCache.size() >= mHardLimit) {
+          return null;
+        } else {
+          return new ValNode(mDefaultLoader.apply(k));
+        }
+      });
+
+      if (cacheEntry == null) {
+        // cache is at hard limit
+        try {
+          if (System.currentTimeMillis() - mLastSizeWarningTime > 60000) {
+            LOG.warn("Cache at hard limit, cache ize = " + mCache.size()
+                + " softLimit = " + mSoftLimit + " hardLimit = " + mHardLimit);
+            mLastSizeWarningTime = System.currentTimeMillis();
+          }
+          Thread.sleep(100);
+          evictIfOverLimit();
+        } catch (InterruptedException e) {
+          throw new RuntimeException(e);
+        }
+        continue;
+      }
+
+      if ((oldCacheEntry != cacheEntry)
+          && cacheEntry.mRefCount.get() > 0) {
+        evictIfOverLimit();
+        return cacheEntry;
+      } else {
+        // refCount went negative, we are evicting this entry
```

Comment:
I think saying "we" is confusing, since this thread is not evicting anything. It is actually being evicted by a different thread.

Reply:
done